src/typecheck/trans_sys.ml

```diff
@@ -0,0 +1,246 @@
+          _duplicate_definition env loc local_name other_loc
+        | None -> (
+          let sub_id, sid_loc, args, inst_loc = get_app_info sub_inst in
+          let sub_inputs, sub_outputs =
+            match M.find_opt sub_id defs with
+            | None ->
+                _cannot_find_system env sid_loc sub_id
+            | Some (`Trans_Sys (input, output, _)) ->
+                (vars input, vars output)
+          in
+          let num_args = List.length args in
+          let params = sub_inputs @ sub_outputs in
+          let num_params = List.length params in
+          if (num_args != num_params) then (
+            _bad_inst_arity env inst_loc sub_id num_params num_args
+          ) ;
+          List.iter2
+            (fun arg param ->
+              let expected_type = Type.T.Var.ty param in
+              parse_ensure env arg expected_type |> ignore
+            )
+            args
+            params ;
+          M.add local_name loc other_subs
+        )
+      )
+      M.empty
+      parent.subs
+    |> ignore
+
+  let parse_def_body ((env, env'), _input, _output, _local) (d: Stmt.sys_def) =
+    parse_condition env d.init ;
+    parse_condition env' d.trans ; 
+    parse_condition env d.inv ;
+    parse_subsystems env d
+
+  let finalize_sys (d : Stmt.sys_def) ((env, _), input, output, local) =
+    Type.check_no_free_wildcards env d.init;
+    Type.check_no_free_wildcards env d.trans;
+    Type.check_no_free_wildcards env d.inv;
+    let input, output, local = vars input, vars output, vars local in
+    (* TODO: review cases of unused variable *)
+    List.iter (Type.check_used_term_var ~kind:`Trans_sys_param env) input ;
+    List.iter (Type.check_used_term_var ~kind:`Trans_sys_param env) output ;
+    List.iter (Type.check_used_term_var ~kind:`Trans_sys_param env) local ;
+    `Sys_def (d.id, input, output, local)
+
+  let parse_def env (d : Stmt.sys_def) =
+    let ssig = parse_sig env d.input d.output d.local in
+    parse_def_body ssig d ;
+    define_sys d.id ssig ;
+    finalize_sys d ssig
+
+  let get_sys_sig env sid =
+    let defs = get_defs env in
+    let id, loc = get_symbol_id_and_loc sid in
+    match M.find_opt id defs with
+    | None -> _cannot_find_system env loc id
+    | Some (`Trans_Sys ssig) -> ssig
+
+  let check_sig (env, env') id sys chk =
+    match chk with
+    | [] -> (
+      List.fold_left
+        (fun (env, env') (id, v, ast) ->
+          let id' = create_primed_id id in
+          let env = Type.add_term_var env id v ast in
+          let env' = Type.add_term_var env' id v ast in
+          let env' = Type.add_term_var env' id' v ast in
+          env, env'
+        )
+        (env, env')
+        sys
+    )
+    | (_, _, a) :: _ -> (
+      let n1 = List.length sys in
+      let n2 = List.length chk in
+      if (n1 != n2) then _bad_arity env (Id id) n1 n2 a ;
+      List.iter2
+        (fun (_, v1, _) (_, v2, a2) ->
+          ensure env a2 (Type.T.of_var v2) (Type.T.Var.ty v1)
```

Comment:
I think that here it would be better to directly check for equality of the types, and raise an error if the equality does not hold.